crates/ruff_linter/src/rules/pycodestyle/rules/too_many_newlines_at_end_of_file.rs

```diff
@@ -48,42 +51,96 @@ impl AlwaysFixableViolation for TooManyNewlinesAtEndOfFile {
 }
 
 /// W391
-pub(crate) fn too_many_newlines_at_end_of_file(diagnostics: &mut Vec<Diagnostic>, tokens: &Tokens) {
-    let mut num_trailing_newlines = 0u32;
-    let mut start: Option<TextSize> = None;
-    let mut end: Option<TextSize> = None;
-
-    // Count the number of trailing newlines.
-    for token in tokens.iter().rev() {
-        match token.kind() {
-            TokenKind::NonLogicalNewline | TokenKind::Newline => {
-                if num_trailing_newlines == 0 {
-                    end = Some(token.end());
-                }
-                start = Some(token.end());
-                num_trailing_newlines += 1;
-            }
-            TokenKind::Dedent => continue,
-            _ => {
+pub(crate) fn too_many_newlines_at_end_of_file(
+    diagnostics: &mut Vec<Diagnostic>,
+    tokens: &Tokens,
+    cell_offsets: Option<&CellOffsets>,
+) {
+    let Some(last_token) = tokens.last() else {
+        return;
+    };
+    let last_textsize = last_token.end();
+
+    let tokens_iter = tokens.iter().rev().peekable();
+
+    let newline_diagnostics = if let Some(cell_offsets) = cell_offsets {
+        let offset_iter = cell_offsets.iter().rev().peekable();
+        collect_trailing_newlines_diagnostics(tokens_iter, offset_iter)
+    } else {
+        // To handle notebooks and ordinary source files uniformly,
+        // we "convert" the source type to a notebook by appending
+        // a newline.
+        let last_textsize = last_textsize + TextSize::from(1);
+        let offset_iter = std::iter::once(&last_textsize).peekable();
+        collect_trailing_newlines_diagnostics(tokens_iter, offset_iter)
```

Comment:
It feels like the only part that needs sharing between notebooks and regular documents is line 98 to 142. Would it make sense to extract this logic into its own function instead of "faking" a notebook for normal documents? 

This would have the advantage that you can move the `cell_offsets.iter().rev` into `collect_trailing_newlines_diagnostics` (you probably want to rename that method), which, in turn, makes the comment about `offset_iter` being reverse obsolete.

I'd also expect that doing this would reduce the diff size.